### PR TITLE
Update DiscoverKubernetesNode with error handling

### PR DESCRIPTION
**Pull request by MichaelKatsoulis**
## What does this PR do?

This PR enhances `DiscoverKubernetesNode` function with error handling. The function, in case kubernetes node is not found will return an error message. The `defaultNode` was changed from `localhost` to the environment variable `NODE_NAME`.
The callers of the function will handle the error message accordingly to their functionality.

1. Kubernetes autodiscover (`NewNodeEventer`, `NewPodEventer`) will fail in case kubernetes Node is not found and this will lead to metricbeat/filebeat/heartbeat failure. This is according to discussions in https://github.com/elastic/beats/issues/23400
2. `add_kubernetes_metadata` processor will fail in case Node is not found and log the error message. However Beats will not fail but events will not be enriched with k8s metadata .
3.  Metricbeat  `NewContainerMetadataEnricher` and `NewResourceMetadataEnricher` will fail in case node is not found. Metricbeat will not fail but events will not be enriched with k8s metadata for kubernetes metricbeat module.
4.  Elastic agent kubernetes dynamic provider will not start in case node is not found. A debug message is logged. Agent will skip this provider.

## Why is it important?
Currently when k8s autodiscover fail due to Node not found  filebeat and metricbeat just proceed with normal operation without shipping any k8s logs and metrics. This is fixed. [Issue](https://github.com/elastic/beats/issues/23400)

Until there is a way in agent/fleet to configure properly Kubernetes dynamic provider and set the Node name correctly, an update was needed in `DiscoverKubernetesNode` to try to find the node name from `NODE_NAME` env var instead returning `localhost` as default. This will allow us to add condition to leverage kubernetes provider for controllermanager and scheduler data streams [PR](https://github.com/elastic/integrations/pull/1324) .

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Related issues

Closes https://github.com/elastic/beats/issues/23400

## How to test this PR locally
In this PR detailed units tests for all cases of `DiscoverKubernetesNode` function have been added. 
Inside beats repo execute `go test -v github.com/elastic/beats/v7/libbeat/common/kubernetes/...`

For manual testing of e.g. `metricbeat`.

1. Following steps of [test metricbeat on k8s](https://github.com/elastic/beats/blob/master/metricbeat/module/kubernetes/_meta/test/docs/README.md) update `metricbeat.yml` in `metricbeat-daemonset-config` ConfigMap with 
```
metricbeat.autodiscover:
      providers:
        - type: kubernetes
          hints.enabled: true
```
From `DaemonSet` env vars comment out 
```
#            - name: NODE_NAME
#              valueFrom:
#                fieldRef:
#                  fieldPath: spec.nodeName
```
Compile and copy metricbeat binary in the pod(the pod sleeps, see instructions) and start metricbeat. 
Notice metricbeat failing.
beat_failure

2. Comment back in `NODE_NAME ` environment variable from daemonset and rerun metricbeat. Notice the debug and info logs. Node is retrieved from the env var.
metricbeat_node_from_env

### CODE REVIEW

File: libbeat/common/kubernetes/util.go [hunk 1]
```diff
@@ -94,54 +93,65 @@ func IsInCluster(kubeconfig string) bool {
 // If host is provided in the config use it directly.
 // If beat is deployed in k8s cluster, use hostname of pod which is pod name to query pod meta for node name.
 // If beat is deployed outside k8s cluster, use machine-id to match against k8s nodes for node name.
-func DiscoverKubernetesNode(log *logp.Logger, host string, inCluster bool, client kubernetes.Interface) (node string) {
+// If node can not be discovered return NODE_NAME env var as default value. In case it is not set return error.
```
Comment (ChrsMark) [inline libbeat/common/kubernetes/util.go:96]:
```suggestion
// If node cannot be discovered, return NODE_NAME env var as default value. In case it is not set return error.
```

File: libbeat/common/kubernetes/util.go [hunk 2]
```diff
@@ -94,54 +93,65 @@ func IsInCluster(kubeconfig string) bool {
...
+			return nodeNameEnv, enrichedError([]error{envError, logerror})
 		}
 		log.Infof("kubernetes: Using pod name %s and namespace %s to discover kubernetes node", podName, ns)
 		pod, err := client.CoreV1().Pods(ns).Get(ctx, podName, metav1.GetOptions{})
 		if err != nil {
-			log.Errorf("kubernetes: Querying for pod failed with error: %+v", err)
-			return defaultNode
+			logerror = fmt.Errorf("kubernetes: Querying for pod failed with error: %+v", err)
+			log.Error(logerror)
+			return nodeNameEnv, enrichedError([]error{envError, logerror})
```
Comment (ChrsMark) [inline libbeat/common/kubernetes/util.go:126]: Maybe `Wrap` can be used here instead of `enrichedError `? See https://github.com/elastic/beats/blob/95626b8f1690344312c0831ab2bdcbccffe4d089/x-pack/metricbeat/module/stan/stats/stats.go#L73

File: libbeat/common/kubernetes/util.go [hunk 3]
```diff
@@ -94,54 +93,65 @@ func IsInCluster(kubeconfig string) bool {
...
+	if nodeNameEnv == "" {
+		envError = errors.New("kubernetes: NODE_NAME environment variable was not set")
 	}
 	ctx := context.TODO()
 	if inCluster {
 		ns, err := InClusterNamespace()
 		if err != nil {
-			log.Errorf("kubernetes: Couldn't get namespace when beat is in cluster with error: %+v", err.Error())
-			return defaultNode
+			logerror = fmt.Errorf("kubernetes: Couldn't get namespace when beat is in cluster with error: %+v", err.Error())
```
Comment (ChrsMark) [inline libbeat/common/kubernetes/util.go:111]: Since there is a lot of `if/else` logic, how about covering these cases with some unit tests? There is a [`fake-go-client`](https://github.com/elastic/beats/blob/49e3857fd412ff95d4374118e6e447aff9ebb619/x-pack/elastic-agent/pkg/composable/providers/kubernetessecrets/kubernetes_secrets_test.go#L25) library you can use to simulate any calls to k8s api.

**Review by ChrsMark — CHANGES_REQUESTED** [review]
Overall the strategy looks good. Please consider some things to add/improve:
1. add a changelog entry
2. add some unit tests
3. mark this for manual testing since the full stack scenarios cannot be covered by just UTs
4. describe  in this PR's description the strategy we will follow per usage of `DiscoverKubernetesNode` so as to have an easy access to this in the future in case we have questions/issues etc.

File: libbeat/common/kubernetes/util.go (hunk 2 above)
Comment (MichaelKatsoulis) [reply libbeat/common/kubernetes/util.go]: Good catch. It does almost the same thing

File: libbeat/common/kubernetes/util.go [hunk 4]
```diff
@@ -94,54 +98,65 @@ func IsInCluster(kubeconfig string) bool {
...
+			return nodeNameEnv, errors.Wrap(envError, logerror.Error())
 		}
 		log.Infof("kubernetes: Using pod name %s and namespace %s to discover kubernetes node", podName, ns)
 		pod, err := client.CoreV1().Pods(ns).Get(ctx, podName, metav1.GetOptions{})
 		if err != nil {
-			log.Errorf("kubernetes: Querying for pod failed with error: %+v", err)
-			return defaultNode
+			logerror = fmt.Errorf("kubernetes: Querying for pod failed with error: %+v", err)
+			log.Error(logerror)
+			return nodeNameEnv, errors.Wrap(envError, logerror.Error())
```
Comment (jsoriano) [inline libbeat/common/kubernetes/util.go:131]: Does this work as expected? I would expect that if some of the methods tried fail, it logs something at the debug level and continue. Here it seems that when any of the methods fail, the whole discovery fails.
I would suggest to refactor this to a sequence of discovery methods tried one after the other. If one works, its result is returned, otherwise the error is logged at the debug level and the function continues.
For simplicity and readability, each one of the discovery methods could possibly go to its own function.

File: libbeat/common/kubernetes/util.go [hunk 5]
```diff
@@ -94,54 +98,65 @@ func IsInCluster(kubeconfig string) bool {
...
+	mid := machineId()
 	if mid == "" {
-		log.Error("kubernetes: Couldn't collect info from any of the files in /etc/machine-id /var/lib/dbus/machine-id")
-		return defaultNode
+		logerror = errors.New("kubernetes: Couldn't collect info from any of the files in /etc/machine-id /var/lib/dbus/machine-id")
+		log.Error(logerror)
+		return nodeNameEnv, errors.Wrap(envError, logerror.Error())
 	}
 
 	nodes, err := client.CoreV1().Nodes().List(ctx, metav1.ListOptions{})
```
Comment (jsoriano) [inline libbeat/common/kubernetes/util.go:144]: Maybe we can add a field selector here, so we don't need to loop over all the nodes in the cluster, something like this (not tested):
```suggestion
	nodes, err := client.CoreV1().Nodes().List(ctx, metav1.ListOptions{FieldSelector: fmt.Sprintf("status.nodeInfo.machineID=%s", mid)})
```

**Review by jsoriano — COMMENTED** [review]
Thanks for working on this! I think this is going to solve some headaches when using beats in kubernetes :slightly_smiling_face: 

I see this is in draft, added only some suggestions.

File: libbeat/common/kubernetes/util.go (hunk 4 above)
Comment (MichaelKatsoulis) [reply libbeat/common/kubernetes/util.go]: Thanks a lot @jsoriano for taking a look. I have done a big refectory of the legacy `DiscoverKubernetesNode`. Regarding your comment I will sum up my opinion:

`DiscoverKubernetesNode` fails in specific cases:
Let's divide them in two categories:
1. `isInCluster` true meaning that kubeconfig exists and is valid. In this scenario `Node` will be retrieved from the pod code is running on. Failing cases are:

-  namespace cannot be found. In that case if namespace is not found we cannot move to next step which is to get the pod in that namespace and get the node it is running on. So it has to fail

-  podname cannot be retrieved from os.Hostname. Same as above without the pod name we cannot find the pod and the node it is running on
-  If namespace and podname exist but getting the actual pod from kubernetes fails. Without the pod it cannot get the node

2. `isInCluster` false. kubeconfig cannot be found or invalid. Code tries to get the `Node` needed via `machineId` of the host it running on. It fails when
a.  MachineId is "". Without a valid value it cannot find a kubernetes node with same MachineId. So it has to fail
b. It fails to list the available kubernetes nodes. Without the nodes it cannot compare the `MachineId` against the node's `MachineId`
c. No Node from the ones retrieved has tha MachineId. 

In all above scenarios discovery won't fail if `NODE_NAME` env var is set.

File: libbeat/common/kubernetes/util.go (hunk 5 above)
Comment (MichaelKatsoulis) [reply libbeat/common/kubernetes/util.go]: I tried but unfortunately that filed is not supported as a field-selector.  This is the list of supported selectors for node : https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/v1/conversion.go#L60

File: CHANGELOG.next.asciidoc [hunk 6]
```diff
@@ -35,6 +35,7 @@ https://github.com/elastic/beats/compare/v7.0.0-alpha2...master[Check the HEAD d
 - Removed beats central management {pull}25696[25696], {issue}23908[23908]
 - MacOSX minimum supported version set to 10.14 {issue}24193{24193}
 - Add daemonset.name in pods controlled by DaemonSets {pull}26808[26808], {issue}25816[25816]
+- Refactor DiscoverKubernetesNode to add error handling and tests  {pull}26947[26947]
```
Comment (sayden) [inline CHANGELOG.next.asciidoc:38]: I think you don't need changelog for this change

File: libbeat/common/kubernetes/util.go [hunk 7]
```diff
@@ -30,10 +30,25 @@ import (
...
+	"github.com/pkg/errors"
+
 	"github.com/elastic/beats/v7/libbeat/logp"
 )
 
-const defaultNode = "localhost"
+const namespaceFilePath = "/var/run/secrets/kubernetes.io/serviceaccount/namespace"
+
+type discoveryUtils struct {
+	eDisc       HostDiscovery
```
Comment (sayden) [inline libbeat/common/kubernetes/util.go:41]: I'd really like a different name for this `eDisc` field. One that clearly states that it's used for host discovery purposes.

File: libbeat/common/kubernetes/util.go [hunk 8]
```diff
@@ -30,10 +30,25 @@ import (
...
+	client      kubernetes.Interface
+	isInCluster bool
+}
+type HostDiscovery interface {
```
Comment (sayden) [inline libbeat/common/kubernetes/util.go:45]: This interface might include a `GetNodeName()` function that will later be used in line 127

File: libbeat/common/kubernetes/util.go [hunk 9]
```diff
@@ -92,60 +107,76 @@ func IsInCluster(kubeconfig string) bool {
...
+
+// discoverKubernetesNode figures out the Kubernetes node to use.
+// If beat is deployed in k8s cluster, use hostname of pod as the pod name to query pod metadata for node name.
 // If beat is deployed outside k8s cluster, use machine-id to match against k8s nodes for node name.
-func DiscoverKubernetesNode(log *logp.Logger, host string, inCluster bool, client kubernetes.Interface) (node string) {
-	if host != "" {
-		log.Infof("kubernetes: Using node %s provided in the config", host)
-		return host
+// If node cannot be discovered, return NODE_NAME env var as default value. In case it is not set return error.
+func (d *discoveryUtils) discoverKubernetesNode(log *logp.Logger) (string, error) {
```
Comment (sayden) [inline libbeat/common/kubernetes/util.go:126]: I think I'd write this method on reverse. I mean to pass in as arguments the stuff that is being used (like `eDisc HostDiscovery` and "compose" `discoveryUtils` with the logger so you don't have to pass down the stack the logger in every method call.

It's even possible that Beats shouldn't be using the logger at this level of the stack, but that will require a deeper analysis.

File: libbeat/common/kubernetes/util.go [hunk 10]
```diff
@@ -158,6 +189,16 @@ func machineID() string {
 	return ""
 }
 
+// GetNamespace gets namespace from serviceaccount when beat is in cluster.
+func (hd *hostDiscovery) GetNamespace() (string, error) {
+	return InClusterNamespace()
```
Comment (sayden) [inline libbeat/common/kubernetes/util.go:194]: Maybe this function should be part of this `HostDiscovery`

**Review by sayden — APPROVED** [review]
Looks good overall. I'd really like the small variable name change I mentioned but it's fine too if not. Let's move this forward at small increments, merge it "today" because the error handling improvement looks really necessary and we can keep improving it in the future. Good work!

File: libbeat/common/kubernetes/util.go (hunk 10 above)
Comment (endorama) [reply libbeat/common/kubernetes/util.go:194]: But the same function is used [in another place](https://sourcegraph.com/github.com/elastic/beats@77e83f975717f16fe21f1602e989e099f451763c/-/blob/x-pack/elastic-agent/pkg/composable/providers/kubernetesleaderelection/kubernetes_leaderelection.go?L71:24), so maybe is due to that.

I agree that `InClusterNamespace` looks a lot like an implementation detail of `hostDiscovery`.

File: libbeat/common/kubernetes/util.go (hunk 7 above)
Comment (endorama) [reply libbeat/common/kubernetes/util.go:41]: I second this, as I find very difficult to understand what is supposed to mean. Some documentation may also works.

**Review by endorama — COMMENTED** [review]
TL;DR: error handling is a must, so okay in moving forward even if I think the `discoverKubernetesNode` still needs refactoring to remove the too many branch conditions present in this implementation.

2¢:
- the `discoverKubernetesNode` method has still the same implementation issues as the previous `DiscoverKubernetesNode` minus the missing error handling. So :+1: for managing error but I think some refactoring is still needed to make it simpler to change and test
- is not clear the role of the `discoveryUtils` struct, as all of it's fields are arguments to the `discoverKubernetesNode` function
- (nit) the `HostDiscovery` interface is neat (as it is easier to test), but I think that it may be helpful if it's implementation `hostDiscovery` was public too, so external code can leverage it.

File: libbeat/common/kubernetes/util.go (hunk 4 above)
Comment (jsoriano) [reply libbeat/common/kubernetes/util.go]: Sorry, first of all, I was wrong about the behaviour, I thought that an error was returned in any case, but this wouldn't happen with `errors.Wrap()` when `envError` is nil. But, someone misreading this code is a bad signal in any case :slightly_smiling_face: There were three errors in the scope, plus the use of wrapping is not the best for readability.

After your refactor this is a bit simpler now, only two errors in the scope, but I still think that we have the opportunity to further improve it.

Some ideas below, please look at them as possible opportunities to improve the readability and maintainability of this code a little bit more, nothing blocker, this is already adding some nice things: error handling and a better fallback method for node discovery.

* You mentioned two clear categories, I think they can be two different functions, one called after the other, this would likely simplify the code.
* It can make sense to try the "machine id" method after the in-cluster one. In that case the use of the environment variable would only be the next thing to try.
* Imagine what you would need to do with current code if you wanted to add another fallback if the environment variable was also not available.
* I think that we should still log something at the debug level if discovery fails and we fallback to the use of `NODE_NAME`. In that case maybe the user has forgotten to explicitly set `host`, or they don't expect the use of `NODE_NAME`, and it may be difficult to investigate as with current code nothing would be logged.
* Not sure about the interface added, I wouldn't expect multiple implementations of this, and it doesn't seem to be leveraged in tests. It looks a bit artificial to me.

File: CHANGELOG.next.asciidoc (hunk 6 above)
Comment (jsoriano) [reply CHANGELOG.next.asciidoc:38]: Well, a changelog is needed, but for the user-facing change, something like "`NODE_NAME` environment variable is used now to discover the local kubernetes node name".

Comment (jsoriano) [reply CHANGELOG.next.asciidoc:38]: And/or "Kubernetes autodiscover fails to start now if it cannot discover the local node name and scope is set to node".

**Comment by MichaelKatsoulis** [comment]
@endorama , @jsoriano , @sayden , @ChrsMark  thank you for taking time to review this and add your comments and suggestions. 

@jsoriano , @endorama 
>>You mentioned two clear categories, I think they can be two different functions, one called after the other, or one as alternative of the other. This would likely simplify the code.
It can make sense to try the "machine id" method after the in-cluster one. In that case the use of the environment variable would only be the next thing to try.

>> the discoverKubernetesNode method has still the same implementation issues as the previous DiscoverKubernetesNode minus the missing error handling. So 👍 for managing error but I think some refactoring is still needed to make it simpler to change and test

You have a valid point. The function is still big with different steps that can split into two separate ones. This way readability can be increased. I went on and made some updates with that in mind.
Another think that I changed to make simpler is the error message returned. In case the `isInCluster` step fails, code will move on with `machineId`. If that also fails `the env var` will be used. If everything fails a generic error message will be returned with some proposal to set the `NODE_NAME` env var.  Meanwhile in all steps that fail a debug message is logged so the user can investigate further. 

 >>I second this, as I find very difficult to understand what is supposed to mean. Some documentation may also works.

I agree. I updated the naming.

>>the HostDiscovery interface is neat (as it is easier to test), but I think that it may be helpful if it's implementation hostDiscovery was public too, so external code can leverage it.

Yes. I made it public and shifted the initialisation of the struct to the callers .

>>Not sure about the interface added, I wouldn't expect multiple implementations of this, and it doesn't seem to be leveraged in tests. It looks a bit artificial to me.

Without the use of interface there is no way(I can find of) to mock the functionality of the utility functions like `GetNamespace, GetPodName, GetMachineID`. The interface is leveraged in the tests. see [link](https://github.com/elastic/beats/blob/c7737c3bbb54c73eb0f2e5b1f6b9446ab94629a2/libbeat/common/kubernetes/util_test.go#L292)

File: libbeat/common/kubernetes/util.go [hunk 11]
```diff
@@ -92,60 +111,89 @@ func IsInCluster(kubeconfig string) bool {
...
+
+	// fallback to environment variable NODE_NAME
+	node = os.Getenv("NODE_NAME")
+	if node != "" {
+		log.Infof("kubernetes: Node %s discovered by NODE_NAME environment variable(fallback method)", node)
+		return node, nil
 	}
 
-	mid := machineID()
+	return "", errors.New("kubernetes: Node could not be discovered with any known method. Consider setting env var NODE_NAME")
```
Comment (jsoriano) [inline libbeat/common/kubernetes/util.go:150]: Nice idea suggesting a solution in the error message!

File: libbeat/common/kubernetes/util_test.go [hunk 12]
```diff
@@ -0,0 +1,316 @@
...
+
+func createMockhu(namespace, podname, machineid string) *mockHostdiscoveryutils {
+	return &mockHostdiscoveryutils{namespace: namespace, podname: podname, machineid: machineid}
+}
+
+type mockHostdiscoveryutils struct {
+	namespace string
+	podname   string
+	machineid string
+}
```
Comment (jsoriano) [inline libbeat/common/kubernetes/util_test.go:300]: Oh, I see the use of the interface :+1:

File: libbeat/common/kubernetes/util.go [hunk 13]
```diff
@@ -30,10 +30,29 @@ import (
...
+// DiscoverKubernetesNodeOpts includes options for discovering kubernetes node
+type DiscoverKubernetesNodeOpts struct {
+	ConfigHost  string
+	Client      kubernetes.Interface
+	IsInCluster bool
+	HostUtils   HostDiscoveryUtils
+}
+
+// Hostdiscoveryutils implements functions of HostDiscoveryUtils interface
+type Hostdiscoveryutils struct{}
```
Comment (jsoriano) [inline libbeat/common/kubernetes/util.go:55]: Nit. Use camel case.
```suggestion
type HostDiscoveryUtils struct{}
```

Or as this is an specific implementation of an interface with the same name, use a name related to the implementation, or just `DefaultDiscoveryUtils`.

File: libbeat/common/kubernetes/util.go [hunk 14]
```diff
@@ -30,10 +30,29 @@ import (
...
-const defaultNode = "localhost"
+const namespaceFilePath = "/var/run/secrets/kubernetes.io/serviceaccount/namespace"
+
+type HostDiscoveryUtils interface {
+	GetNamespace() (string, error)
+	GetPodName() (string, error)
+	GetMachineID() string
+}
+
+// DiscoverKubernetesNodeOpts includes options for discovering kubernetes node
```
Comment (jsoriano) [inline libbeat/common/kubernetes/util.go:46]: Nit. This includes more than options :slightly_smiling_face: it also includes a fully fleshed Kubernetes client and an implementation of discovery utils.

Maybe this is a signal that `DiscoveryKubernetesNode()` could be a member of an object containing these fields, and maybe receive the options as parameter.

File: libbeat/autodiscover/providers/kubernetes/pod.go [hunk 15]
```diff
@@ -63,7 +63,11 @@ func NewPodEventer(uuid uuid.UUID, cfg *common.Config, client k8s.Interface, pub
 	// Ensure that node is set correctly whenever the scope is set to "node". Make sure that node is empty
 	// when cluster scope is enforced.
 	if config.Scope == "node" {
-		config.Node = kubernetes.DiscoverKubernetesNode(logger, config.Node, kubernetes.IsInCluster(config.KubeConfig), client)
+		nd := &kubernetes.DiscoverKubernetesNodeOpts{ConfigHost: config.Node, Client: client, IsInCluster: kubernetes.IsInCluster(config.KubeConfig), HostUtils: &kubernetes.Hostdiscoveryutils{}}
```
Comment (jsoriano) [inline libbeat/autodiscover/providers/kubernetes/pod.go:66]: Nit. Consider splitting this in multiple lines so it is easier to read.
```suggestion
		nd := &kubernetes.DiscoverKubernetesNodeOpts{
		    ConfigHost: config.Node,
		    Client: client,
		    IsInCluster: kubernetes.IsInCluster(config.KubeConfig),
		    HostUtils: &kubernetes.Hostdiscoveryutils{},
		}
```

File: libbeat/common/kubernetes/util_test.go [hunk 16]
```diff
@@ -0,0 +1,316 @@
...
+				if err != nil {
+					t.Fatal(err)
+				}
+				defer func() {
+					err := client.CoreV1().Nodes().Delete(context.Background(), "worker-2", metav1.DeleteOptions{})
+					if err != nil {
+						t.Fatalf("failed to delete k8s node: %v", err)
+					}
+				}()
+			}
```
Comment (jsoriano) [inline libbeat/common/kubernetes/util_test.go:217]: This conditional code based on the test name is error prone, a change in the title would unexpectedly change the result of the test.

If some initialization code needs to be run for some test cases, consider adding a `init(*testing.T, kubernetes.client)` function to the test cases struct, so you can do here something like this:
```suggestion
			if test.init != nil {
			    test.init(t, client)
			}
```

**Review by jsoriano — COMMENTED** [review]
Good job here :+1: Added some nitpicking but nothing very important.

File: libbeat/common/kubernetes/util_test.go (hunk 16 above)
Comment (jsoriano) [reply libbeat/common/kubernetes/util_test.go:217]: @MichaelKatsoulis to be clear with my review, this would be the only thing I would like to see changed before merging, the rest of the comments can be seen as suggestions and no blockers at all :slightly_smiling_face:

Comment (MichaelKatsoulis) [reply libbeat/common/kubernetes/util_test.go]: Thanks @jsoriano. Great suggestion! I updated it.

File: libbeat/common/kubernetes/util.go (hunk 14 above)
Comment (MichaelKatsoulis) [reply libbeat/common/kubernetes/util.go]: I changed the name to DiscoverKubernetesNodeParams which is more clear to me

File: libbeat/common/kubernetes/util_test.go [hunk 17]
```diff
@@ -0,0 +1,323 @@
...
+					if err != nil {
+						t.Fatalf("failed to delete k8s pod: %v", err)
+					}
+				}()
+				defer func() {
+					err := client.CoreV1().Nodes().Delete(context.Background(), "worker-2", metav1.DeleteOptions{})
+					if err != nil {
+						t.Fatalf("failed to delete k8s node: %v", err)
+					}
+				}()
```
Comment (jsoriano) [inline libbeat/common/kubernetes/util_test.go:209]: Init functions could define their own cleanups with [`t.Cleanup()`](https://pkg.go.dev/testing#T.Cleanup).

**Review by ChrsMark — APPROVED** [review]

File: libbeat/common/kubernetes/util_test.go (hunk 17 above)
Comment (MichaelKatsoulis) [reply libbeat/common/kubernetes/util_test.go]: Updated!

**Review by jsoriano — APPROVED** [review]
:+1:

**Comment by ChrsMark** [comment]
🚀
